oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/AlarmRulesWatcher.java

```diff
@@ -54,26 +57,48 @@ public class AlarmRulesWatcher extends ConfigChangeWatcher {
     private volatile Map<String, Set<String>> exprMetricsMap;
     private volatile Rules rules;
     private volatile String settingsString;
+    private final ReentrantLock lock;
+    private final AtomicBoolean notifiedByDynamicConfig;
 
     public AlarmRulesWatcher(Rules defaultRules, ModuleProvider provider) {
         super(AlarmModule.NAME, provider, "alarm-settings");
         this.runningContext = new HashMap<>();
         this.alarmRuleRunningRuleMap = new HashMap<>();
         this.exprMetricsMap = new HashMap<>();
         this.settingsString = null;
+        this.lock = new ReentrantLock();
+        this.notifiedByDynamicConfig = new AtomicBoolean(false);
         notify(defaultRules);
     }
 
     @Override
     public void notify(ConfigChangeEvent value) {
-        if (value.getEventType().equals(EventType.DELETE)) {
-            settingsString = null;
-            notify(new Rules());
-        } else {
-            settingsString = value.getNewValue();
-            RulesReader rulesReader = new RulesReader(new StringReader(value.getNewValue()));
-            Rules rules = rulesReader.readRules();
-            notify(rules);
+        lock.lock();
+        try {
+            if (value.getEventType().equals(EventType.DELETE)) {
+                settingsString = null;
+                notify(new Rules());
+            } else {
+                settingsString = value.getNewValue();
+                RulesReader rulesReader = new RulesReader(new StringReader(value.getNewValue()));
+                Rules rules = rulesReader.readRules();
+                notify(rules);
+            }
+            notifiedByDynamicConfig.set(true);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public void notifyBeyondDynamicConfig(Rules newRules) {
```

Comment:
The word beyond seems a little confused. How about initConfig directly?

Reply:
> The word beyond seems a little confused. How about initConfig directly?

Should I use the word 'notifyInitConfig'

Reply:
I would say no need to add notify.

Reply:
> I would say no need to add notify.

Okay, it has been changed.